module/zfs/spa_errlog.c

```diff
@@ -128,6 +225,289 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
 	mutex_exit(&spa->spa_errlist_lock);
 }
 
+#ifdef _KERNEL
+static int
+find_block_txg(dsl_dataset_t *ds, zbookmark_err_phys_t *zep,
+    uint64_t *birth_txg)
+{
+	objset_t *os;
+	if (dmu_objset_from_ds(ds, &os) != 0) {
+		return (SET_ERROR(ENOENT));
+	}
+
+	dnode_t *dn;
+	blkptr_t bp;
+
+	if (dnode_hold(os, zep->zb_object, FTAG, &dn) != 0) {
+		return (SET_ERROR(ENOENT));
+	}
+
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	int err = dbuf_dnode_findbp(dn, zep->zb_level, zep->zb_blkid, &bp, NULL,
+	    NULL);
+
+	if (err != 0 || BP_IS_HOLE(&bp)) {
+		rw_exit(&dn->dn_struct_rwlock);
+		dnode_rele(dn, FTAG);
+		return (SET_ERROR(ENOENT));
+	}
+
+	*birth_txg = bp.blk_birth;
+	rw_exit(&dn->dn_struct_rwlock);
+	dnode_rele(dn, FTAG);
+	return (0);
+}
+
```

Comment:
This is a pretty big function.  Can you add a comment describing this it does, what its arguments are, and what the return value is?